File: php/src/TransactionRepository.php
<?php


namespace Hantsy\TransactionAnalyser;


use Brick\DateTime\LocalDateTime;
use DateTime;

class TransactionRepository
{
    /**
     * @var TransactionLoader
     */
    private TransactionLoader $loader;

    private array $transactions;

    public function __construct(TransactionLoader $loader)
    {
        $this->loader = $loader;
        $this->transactions = $loader->load();
    }

    public function queryByMerchantAndDateRange(
        string $merchant,
        string $fromDate,
        string $toDate
    ): array
    {
        //var_dump($transactions);
        // generating report.
        $reversalTransactions = array_filter($this->transactions,
            function ($v, $k) {
                return $v->getType() == TransactionType::REVERSAL;
            },
            ARRAY_FILTER_USE_BOTH
        );

//        echo "\$reversalTransactions:" . PHP_EOL;
//        var_dump($reversalTransactions);

        $reversalRelatedTransactionIds = array_map(
            function ($v) {
                return $v->getRelatedTransactionId();
            },
            $reversalTransactions
        );

//        echo "\$reversalRelatedTransactionIds:" . PHP_EOL;
//        var_dump($reversalRelatedTransactionIds);

        return array_filter($this->transactions,
            callback: function ($v, $k) use ($reversalRelatedTransactionIds, $toDate, $fromDate, $merchant) {
                return $v->getMerchantName() == $merchant
                    && $v->getTransactedAt()->isAfter(LocalDateTime::fromDateTime(DateTime::createFromFormat('d/m/Y H:i:s', $fromDate)))
                    && $v->getTransactedAt()->isBefore(LocalDateTime::fromDateTime(DateTime::createFromFormat('d/m/Y H:i:s', $toDate)))
                    && $v->getType() == TransactionType::PAYMENT
                    && !in_array($v->getId(), $reversalRelatedTransactionIds);
            },
            mode: ARRAY_FILTER_USE_BOTH
        );
    }
}

File: csharp/TransactionRepository.cs
using System;
using System.Linq;

namespace TransactionAnalyzer
{
    class TransactionRepository
    {
        private readonly Transaction[] _data;

        public TransactionRepository(TransactionLoader loader)
        {
            this._data = loader.Load();
        }

        public Transaction[] QueryByMerchantAndDateRange(
            string merchant,
            DateTime fromDate,
            DateTime toDate
        )
        {
            var reversalRelatedIds = this._data
                .Where(s => s.Type == TransactionType.REVERSAL)
                .Select(s => s.RelatedTransactionId);

           return this._data.Where(s => s.MerchantName.Equals(merchant)
                                  && s.TransactedAt > fromDate
                                  && s.TransactedAt < toDate
                                  && s.Type == TransactionType.PAYMENT
                                  && !reversalRelatedIds.Contains(s.Id)
            ).ToArray();
        }
    }
    
}